CONTRACTS: ensure at most one predicate per pointer

remi-delmas-3000 opened the pull request:
Please ignore the first commit, it is from https://github.com/diffblue/cbmc/pull/8576 and will disappear when rebasing.


Fixes a discrepancy between assert and assume behaviour for pointer predicates, by ensuring
that at most one predicate occurrence can be established at all times, in both assume and assert
contexts.

Before, `is_fresh(p, n) && is_fresh(p, n)` would fail in assert contexts but pass in assume contexts
by allocating twice.


- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ x White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Refresh memory predicate documentation, add paragraph about using pointer predicates in disjunctions.

remi-delmas-3000 commented:
@tautschnig this one is also ready to merge

tautschnig commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -1252,6 +1265,14 @@ __CPROVER_HIDE:;
     {
       return 0;
     }
+    // __CPROVER_assert(
+    //   write_set->linked_ptr_pred_ctx->ptr_pred != ptr1,
+    //   "__CPROVER_pointer_equals does not conflict with other pointer "
+    //   "predicate");
+    // write_set->linked_ptr_pred_ctx->ptr_pred =
+    //   __VERIFIER_nondet___CPROVER_bool()
+    //     ? ptr1
+    //     : write_set->linked_ptr_pred_ctx->ptr_pred;
```
What to take away from this commented-out part?

remi-delmas-3000 replied:
Thanks for catching this. I need to fix that and add more tests for predicates used in ensures clauses in contract replacement mode.

Commit:
CONTRACTS: at most one predicate occurence per pointer

Ensures that at most one pointer predicate is positively
asserted/assumed per pointer.
 Before, `is_fresh(p, n) && is_fresh(p, n)` would
fail in assertion contexts as expected but pass in
assumption contexts by allocating twice.

remi-delmas-3000 commented:
@tautschnig test suite now check every combination of conflicting predicates in both requires and ensures clauses for both contract checking and replacement, we're good to go
